client/services/ReinforcementGeneratorService.js

```diff
@@ -0,0 +1,101 @@
+    return {
+      getReinforcement: function(task, codeEvalResult) {
+
+        var testsKeyedByTag = function(tests) {
+          var splitTests = {};
+          for (var i = 0; i < tests.length; ++i) {
+            var test = tests[i];
+            test.originalIndex = i;
+            var testTag = test.getTag();
+            if (splitTests.hasOwnProperty(testTag)) {
+              splitTests[testTag].push(test);
+            } else {
+              splitTests[testTag] = [test];
+            }
+          }
+          return splitTests;
+        };
+
+        var previousSnapshot = TranscriptService.getPreviousSnapshot();
+        var previousReinforcement = null;
+        if (previousSnapshot !== null) {
+          previousReinforcement = previousSnapshot.getReinforcement();
+        }
+
+        var reinforcement = ReinforcementObjectFactory.create(task);
+
+        // Copy the previous reinforcement object if we are on the same task.
+        if (previousReinforcement !== null &&
+            task === previousReinforcement.getTask()) {
+          for (var tag in previousReinforcement.getPassedList()) {
+            reinforcement.addToPassedList(tag,
+              previousReinforcement.getPassedList()[tag]);
+          }
+          for (var pastCase in previousReinforcement.getPastFailsList()) {
+            reinforcement.addToPastFailsList(pastCase,
+              previousReinforcement.getPastFailsList()[pastCase]);
+          }
+        }
+
+        // Go through correctness tests to update reinforcement data.
+        var correctnessTests = testsKeyedByTag(task.getCorrectnessTests());
+        var observedOutputs = codeEvalResult.getCorrectnessTestResults();
+        // Get results only for the last task.
+        observedOutputs = observedOutputs[observedOutputs.length - 1];
+        var failedCaseSeenOverall = false;
+        for (var testTag in correctnessTests) {
+          var failedCaseSeenInTag = false;
+          for (var testIdx = 0; testIdx < correctnessTests[testTag].length;
+              ++testIdx) {
+            var test = correctnessTests[testTag][testIdx];
+            var observedOutput = observedOutputs[test.originalIndex];
+            if (test.matchesOutput(observedOutput)) {
+              if (test.getInput() in reinforcement.getPastFailsList()) {
+                reinforcement.addToPastFailsList(test.getInput(), true);
+              }
+            } else if (!failedCaseSeenInTag) {
+              // We want to display only 1 new failed case among all tasks.
+              if (!failedCaseSeenOverall) {
+                reinforcement.addToPastFailsList(test.getInput(), false);
+                failedCaseSeenOverall = true;
+              }
+              failedCaseSeenInTag = true;
+            }
+          }
+          if (failedCaseSeenInTag) {
+            if (testTag in reinforcement.getPassedList()) {
+              reinforcement.addToPassedList(testTag, false);
```

Comment:
I'm starting to think that "passed list" is a bit of a misnomer here. What are your two lists tracking, exactly? Can the names better reflect their contents?